Remove antiquated msmtp and ntfy images

jetchirag opened the pull request:
Following the discussion on #130 and #144, I've made the necessary changes to discontinue the msmtp and ntfy flavors of the Docker image in this pull request.

I've also moved the base flavor to top level of the repo.

witten commented:
I'm not really a dev on this project, but just chiming in: This looks great to me!

m3nu commented:
Just tried this from building to using and I think it's a great start. (I never used this image before). Here my thoughts and ideas. I know that you took some of this from the previous version, so this is just general discussion, not this PR specifically. Some things are also more ideas and if there are arguments against them, we should discuss them.

### README
- The section 'Starting and stopping containers from hooks' is a very special use case that should be further down with examples on how to mount the Docker socket.
- "Example run command": This is the most important part. How can I quickly set up my backups. Since this needs some folders and setup outside the container, we should include those commands too. E.g.
  - Create required folders: `mkdir -p ./data/{borgmatic.d,.cache,.config,...}
  - Adjust or copy `borgmatic.yaml`
  - Init repository
- Environment: All supported env vars should be listed (as table?) with default value and if they are required. 

#### README structure
High level sections in the README could be:

1. Description: What this does? Run Borg and Borgmatic in a container. Optionally the Cron daemon.
2. Usage: Different ways to use this image
   1. The minimal shell commands to run Borgmatic as binary only. Similar to *Example run command*. `mkdir ...; docker run ... borg init...`
   2. Running as daemon: Still using `docker run` here, but keep the it running for continuous backups
   3. Running with `docker compose`. For more structured deployment. Links to the `docker-compose.yml` file and `.env.template` file and shows how to adjust.
3. Volumes: Explain which folders can be mounted. Best as table
4. Environment Variables: Explain which env vars can be passt. Best as table.  Should mention that all Borg/Borgmatic env vars will also work and link to them.
5. Additional Uses
  - Stopping containers with hooks/including Docker socket
  - FUSE

### Usage as binary or daemon
It took me a while to figure out that this image will by default start a Cron daemon and not do anything for a while. I feel like using this image as `borgmatic` binary is a second use case to cover. E.g. as described [here](https://500.keboola.com/run-any-binary-in-a-container-like-it-exists-on-your-computer/).

Question is if we let users use their own Cron (like a Systemd timer) or manual backups only. I feel forcing this cron daemon is limiting the usefulness. Both uses, as daemon and as binary are useful.

How about explaining this Cron thing better and adjusting the entryscript to allow usage as binary IF a subcommand is provided. E.g. this will run `borgmatic list`:

```
$ docker run \
  --rm -it \
  -v /home/manu/Workspace:/mnt/source:ro,Z \
  -v ./data/borgmatic.d:/etc/borgmatic.d/:Z \
  -v ./data/.config/borg:/root/.config/borg:Z \
  -v ~/.ssh:/root/.ssh:Z \
  ghcr.io/borgmatic-collective/borgmatic \
  list
```
Issue here is of course that we have so many mounts to pass each time, so the command is a bit long. I would still offer this case as example.

The entryscript would need to be adjusted to check for commands to be specified. If none are set, run the Cron daemon. If they are set and aren't `bash` or `sh`, run them as `borgmatic` subcommand?

### Volumes
- Never seen so many volumes. Hope this can be simplified.
- `borgmatic.d` will only contain a single `borgmatic.yaml` file? So why not mount only this file so it's easier to work with on the host?
- `VOLUME_BORGMATIC_STATE` var is missing in the `.env.template` file. Some of those volumes can be merged maybe to simplify it?


### Building
- `.github/workflows/build_*`: Those files look almost the same. Sure we need both?
- `Dockerfile`, `COPY --link requirements.txt /`. I removed `--link` here since it's buildkit specific and didn't work with Podman which I use.
- I was never able to stop the container gracefully. Seems like this cron software needs SIGKILL

m3nu commented:
(Since my suggestion mostly restructures the README, it could be a separate PR)

jetchirag commented:
@m3nu Thank you for all the suggestions! I've writing a new readme file, section by section, but I expect it to take some time since I need to understand some of the points myself so I can better explain them in the guide.

For VOLUME_BORGMATIC_STATE, see #197 

As for the PR, I'm thinking if I should use this one for these changes?

m3nu commented:
Thanks for taking this into account. I think it will greatly help people use this role. That's based on my experience with running a Docker hosting service and integrating about 60 different apps and the support provided for BorgBase.

If you want, you can adjust this PR. Since it's limited to the README. For other things, maybe we add a new PR? E.g. making it easier to use the image as binary.

Also, for the next milestone, I don't see too much to do. The image is already quite slim now. So making it multi-stage and removing a few deps will be quick. We can spend the other time on usability.

modem7 commented:
Hey guys - thought I'd post quickly as I'm not ignoring this. 

Work is slamming me, will try and get a review done on the weekend if that's alright!

jetchirag commented:
@m3nu I've pushed just changes entry file changes. Can you check if that's what you were referring to? I'll explain the usage in Readme.

I was also wondering if that's how images are generally built since I haven't much many. Since it could be achieved before by passing whole command `borgmatic list` instead of just subcommand. And if there are any cons of this approach.

m3nu commented:
exec sounds like a good solution then. I'm sure other projects have already solved this issue too.

I'd keep Dockerfiles simple to avoid being dependent on specific versions. Already tried buildkit in podman, but it didn't work with e.g. docker-compose.

crontab.txt is only for this image. It's not from Borgmatic. Could the cron schedule be an env var?

jetchirag commented:
Makes sense. I'll remove `--link`.

> crontab.txt is only for this image. It's not from Borgmatic. Could the cron schedule be an env var?

It can be. I can have it create cronjob file in `entry.sh` with passed variable or default to `0 1 * * *`. Follow up question, is one a day enough for default value?

m3nu commented:
Yes, once a day is fine as default. Daily backups are most common. Especially with more data, like a few TB.

m3nu commented:
Was thinking about the "as binary" thing a bit more: Maybe it's not too clear what happens, if we treat some commands as Borgmatic commands and somehow redirect them.

Check out this Restic Docker setup: https://github.com/djmaze/resticker

They have some good concepts:

- I like how they use env vars for Cron
- And how they show running any command via docker-compose ([here](https://github.com/djmaze/resticker))
- Since we have many volumes, having this config in Docker-compose makes sense. But README should also show how to do it without.
- I like the `RUN_ON_STARTUP` option to do a backup right away.
- They also use exec as your suggested. And the [entryscript](https://github.com/djmaze/resticker/blob/master/entrypoint) knows about the subcommand.

rallisf1 commented:
> > borgmatic.d will only contain a single borgmatic.yaml file? So why not mount only this file so it's easier to work with on the host?
> 
> Seems to contain two files, `config.yaml` and `crontab.txt`

I have tons of files in borgmatic.d as I am backing up multiple projects in the same repo by using [archive_name_format](https://torsion.org/borgmatic/docs/how-to/make-per-application-backups/#archive-naming)

m3nu commented:
Noted, then let's see if any other mounts could be optional or grouped.

jetchirag commented:
> And how they show running any command via docker-compose

Wrote about that [here](And how they show running any command via docker-compose). I did however run into [a problem I couldn't find a fix to.](https://github.com/borgmatic-collective/docker-borgmatic/pull/240/commits/a677dfbbd9799b62a6f001a1e19783eecad8a916#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5R103)

> Since we have many volumes, having this config in Docker-compose makes sense. But README should also show how to do it without

I added a table previously. Do we need anything else?

> They also use exec as your suggested. And the [entryscript](https://github.com/djmaze/resticker/blob/master/entrypoint) knows about the subcommand.

It seems to only know about `unlock` command otherwise it's also executing `exec restic "$@"` as my earlier commit.  It is however verifying the repo. Should we add something like that too?

Also added the other two variables you mentioned.

modem7 reviewed (approved):
From what I can see, this looks good - I have not tested the code changes, however.

m3nu commented:
Yeah, much better than before and you went the extra mile by expanding the binary use case even further.

Other thoughts:

- I suggest moving the [*Example Run Command*](https://github.com/jetchirag/docker-borgmatic/tree/master#example-run-command) to the top and call it like *Minimal Example Command* or something. So users can quickly try it without doing much setup. This could include `mkdir -p ...` to create the folders needed for mounting. The stuff about security can stay under *Advanced*.
- Under [Prerequisites](https://github.com/jetchirag/docker-borgmatic/tree/master#prerequisites), you could add a link to Docker and maybe Podman as alternative. I often see this for people who haven't used Docker much before. Same for Docker-compose further down.
- You are using `MOUNT_FLAGS` as placeholder, which is fine. Maybe you add a link to the *Volumes* section to look those up. Moving the example with all volumes to the top will help too.

Rest looks good. Also good to see that @modem7 who originally came up with much of this likes it too. Thanks for working on this and reviewing, everyone! 🙏

jetchirag commented:
Restructured it and I think all your points are met now.

m3nu commented:
Looks good! Will run it locally and see if anything stands out.

Else, @witten can get ready to merge this. And deal with the merge conflicts I see here.

witten commented:
@modem7, if you have the time, it'd be great if you could have the final say on this one. Otherwise, let me know and I can do it. Thanks.

grantbevis commented on Dockerfile:
```diff
@@ -33,9 +33,9 @@ RUN apk add --update --no-cache \
 RUN borgmatic --bash-completion > /usr/share/bash-completion/completions/borgmatic && echo "source /etc/profile.d/bash_completion.sh" > /root/.bashrc
 
-CMD ["/entry.sh"]
+ENTRYPOINT ["/entry.sh"]
```
The reason this is `CMD` and not `ENTRYPOINT` is so people can `docker run -it --rm b3vis/docker-borgmatic bash` and get a shell if needed

grantbevis replied:
I see this is to support the binary use case - resolving

grantbevis commented:
Looks good to me @jetchirag 👍🏽

witten commented:
I think a Grant review is sufficient for now! Thanks.

witten reviewed (approved):
I did a merge from master.

jetchirag commented:
Thank you for solving merge conflicts.

I couldn't find a solution to this. Can someone take a look?

https://github.com/borgmatic-collective/docker-borgmatic/blob/638c2d647fb6a599574e58cd21dddafd9cb57e6b/README.md?plain=1#L170-L174
